src/math/Matrix3.d.ts

```diff
@@ -125,6 +131,7 @@ export class Matrix3 implements Matrix {
 	 * @return The created or provided array.
 	 */
 	toArray( array?: number[], offset?: number ): number[];
+	toArray( array?: Matrix3tuple, offset?: 0 ): Matrix3tuple;
```

Comment:
Why `offset?: 0`?

Reply:
That's a good question, thanks.  The special case I've inserted does indeed depend upon the offset being zero.  If the offset changes, then the length of the array returned changes, so it is no longer a Matrix3tuple, or whatever, so TypeScript should use the existing more general definition of toArray in those circumstances.

Reply:
I believe the argument should remain a `number[]` here — it could have any length — but the return type could be a tuple.

Reply:
That's a good thought, but I can change the return type by changing the length of the array in that parameter.

![image](https://user-images.githubusercontent.com/1577335/98027887-3c48b600-1e05-11eb-98f7-da96dde01031.png)

Reply:
Oh, the actual return type of the method can actually be `number[]` then, yes. You could handle this with generics:

```typescript
toArray<T = number[] | vec2>(target: T): T;
```

That said, I'm not sure generics would actually make type-checking work any better here, compared to just letting the user cast the result:

```typescript
const result = vector.toArray([]) as vec2;
```

Reply:
Yes, casting the result is a valid workaround for the original bug anyway.

That's clever to use generics to try and combine the two alternative definitions - the original more general one, and my more specific one - into a single definition.  It doesn't work because of how the value given for the offset parameter changes the return type.  Code utilising generics is sometimes very powerful, but it is harder to read.  Remember that the various definitions of `toArray()` do appear in the context-senstive help while you are entering its parameters via an IDE such as Visual Studio Code.

Reply:
So with two function overloads available, the compiler chooses?

```typescript
// loose definition
const tmp = [];
vector.toArray(tmp); // → number[], matches loose definition
vector.toArray(tmp as vec3, 100); // number[], matches loose definition

// strict definition
const tmp = [] as vec3;
vector.toArray(tmp); // → vec3, matches subclass definition
vector.toArray(tmp, 0); // → vec3, matches subclass definition
```

I think I see now. 👍

Reply:
Let me know if there's anything further I can be of assistance with.  I believe I have answered all your questions.  I appreciate it's difficult to find the time to sit down and play with my new definition, and I'm not trying to hurry you.  It's entirely up to you guys what you decide and when.

The idea here is simply to provide more detailed type information for the most common usages of `toArray`.  Specifically, in order to use the spread operator (`...`) with methods such as `translate` which require a specific number of arguments the compiler needs to know, not just that `toArray` returns an array, but also that the length of that returned array matches the number of arguments.

As @donmccurdy has already figured out, this change is non-breaking because the compiler is happy so long as it can find one definition of `toArray` that matches, so adding an additionally definition shouldn't be able to take away from what's already there.  Of course that's not enough on it's own, you also need to check that the definition is correct, because if it transpired it wasn't and you had to correct it, that would almost certainly break code compiled against the definitions I have added here.  I don't think it's too hard to see that the new definitions simply describe a particular case of what `toArray` already does, and I wish you well with coming to your decision about this.